test(snapshot/bdd/create): add feature file for snapshot creation

tiagolobocastro opened the pull request.

tiagolobocastro committed:
test(snapshot/bdd/create): add feature file for snapshot creation

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

dsharma-dc commented on tests/bdd/features/snapshot/create/feature.feature:
```diff
@@ -0,0 +1,78 @@
+    And the snapshot source size should be equal to the volume size
+    And the snapshot size should be equal to the volume size
+    And the snapshot timestamp should be in the RFC 3339 format
+    And the snapshot timestamp should be within 1 minute of creation (UTC)
```
We don't really need to mention timezone UTC here specifically, as we are talking about a delta.

dsharma-dc replied:
What do we consider "creation" time here? The timestamp at which test will call put_volume_snapshot?

tiagolobocastro replied:
> We don't really need to mention timezone UTC here specifically, as we are talking about a delta.

We might do because if your computer is not on UTC then you can't just compare time as the snapshot time is in utc?

> What do we consider "creation" time here? The timestamp at which test will call put_volume_snapshot?

It doesn't matter too much if it's right before we issue the call or after, 1 minute will be a large enough slack for work for both cases

dsharma-dc commented on tests/bdd/features/snapshot/create/feature.feature:
```diff
@@ -0,0 +1,78 @@
+    And volume is in faulted state
+    When we try to create a snapshot for the volume
+    Then the snapshot creation should fail with preconditions failed
+    And no snapshot should be listed
```
Rather `the snapshot uuid used in create snapshot request shouldn't get listed` , because we may have possibly existing snapshots for the replica.

dsharma-dc commented on tests/bdd/features/snapshot/create/feature.feature:
```diff
@@ -0,0 +1,78 @@
+    When a snapshot is created for the volume
+    When the volume is deleted
+    And we retry the same snapshot creation
+    Then it should fail with already exists
```
This one looks confusing to me. Are we testing that `same snapshot/id creation should fail with eexist` i.e. the last scenario down in this file, or we are testing that `any snapshot creation after volume deletion should fail` ?

tiagolobocastro replied:
we're testing that it fails with eexists even when the volume has been deleted

Abhinandan-Purkait commented on tests/bdd/features/snapshot/create/feature.feature:
```diff
@@ -0,0 +1,78 @@
+  Scenario: Existence of snapshot after node restart
+    Given we have a single replica volume
+    And a successfully created snapshot
+    When we stop the node hosting the snapshot
```
Instead of stopping and starting can we just restart?

Abhinandan-Purkait replied:
And if we want to stop and start then probably should we kill and start, i.e ungracefully.

tiagolobocastro replied:
why does it matter whether we stop or kill here?

Abhinandan-Purkait replied:
Stop does things gracefully, which might not be the case always.

tiagolobocastro replied:
Vice versa :) upgrades etc will use stop

tiagolobocastro replied:
Maybe we could use outline and test both

Abhinandan-Purkait replied:
Right.

Abhinandan-Purkait commented on tests/bdd/features/snapshot/create/feature.feature:
```diff
@@ -0,0 +1,78 @@
+  Scenario: Retrying snapshot creation until it succeeds
+    Given we have a single replica volume
+    And the volume is published on node A
+    And the node A is brought down
```
pause rather than stop ? Otherwise we might hit precondition? :-p

tiagolobocastro replied:
Right :D

tiagolobocastro committed:
fix(snapshot): make timestamp optional and return correct errors

We might not have a state from dataplane so state timestamp should be optional.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(snapshot/bdd/create): add create snapshot bdd test impl

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on control-plane/rest/openapi-specs/v0_api_spec.yaml:
```diff
@@ -3305,7 +3308,6 @@ components:
         - uuid
         - size
         - source_volume
-        - timestamp
```
Do we even need the timestamp for VolumeSnapshotState? Csi will consume from metadata. And in state we have the individual replica snapshot state time stamp.

tiagolobocastro replied:
hmm we might not even need it yes

Abhinandan-Purkait commented on tests/bdd/features/snapshot/create/test_feature.py:
```diff
@@ -0,0 +1,527 @@
+    try:
+        pytest.snapshot = snapshot_create_request()
+    except openapi.exceptions.ApiException as e:
+        print(ApiClient.snapshots_api().get_volumes_snapshots())
```
Don't need the print?

tiagolobocastro replied:
No, it's from debugging :D

tiagolobocastro committed:
fix(snapshots/ready): mark snapshots as ready if they're in the Created phase

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
chore(bdd): speed up existing tests

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge
